### ISSUE TITLE
Revise NiFTi specifications in Sanitize interface

### ISSUE THREAD

**Issue by oesteban** [poldracklab/niworkflows#222]
Revise the truth table we are using, and the messages we give making sure that:

1. [Valid codes](https://nifti.nimh.nih.gov/nifti-1/documentation/nifti1fields/nifti1fields_pages/qsform_brief_usage):
> The qform_code should be set to either NIFTI_XFORM_UNKNOWN or NIFTI_XFORM_SCANNER_ANAT.

and

> The sform code should be set to either NIFTI_XFORM_UNKNOWN, NIFTI_XFORM_ALIGNED_ANAT, NIFTI_XFORM_TALAIRACH or NIFTI_XFORM_MNI_152.

So `qform_code = [0|1]` and `sform_code = [0|2|3|4]`. These limitations should be considered in the table.

If both qform and sform code make sense and are apparently valid, also check:
> The qform and sform should never specify coordinates with different handedness (i.e. have determinants of different signs). Otherwise it is not possible to tell left from right.

2. ``qform_code == 0``: this is old ANALYZE 7.5 compatibility mode and it is discouraged. Warn that header will be updated. If ``sform_code == 0`` it is calculated using the pixel dimensions, assuming LAS orientation of ANALYZE 7.5 (nibabel's best affine, I'd guess). If ``sform_code > 0``, then we use that sform.

2. Add a `force_nifti_method_2` to the inputs. This would indicate that the file should not specify an sform (because it is "raw" data and no alignment to anatomical or atlas is done). However, in those situations where `qform_code == 0` or it is invalid, then we would make use of the sform (if set, valid and rigid -not affine-).

3. If both sform and qform codes are >0, and both matrices valid, then we look to `force_nifti_method_2`. If ``sform_code > 2`` and`force_nifti_method_1 is True`, then the sform is cleaned up (identity and set code 0). We probably want to have a new input `afni_compat`. If `afni_compat is True` then the qform is copied into the sform.

4. We would need to think of the special case of ``sform_code == 2``, or make the boolean mean "aligned to anything (atlas, subject)". In this case we would be dropping the sform that bbregister and flirt may use (although I don't think it will affect at all since we have an initialization step).

The former behavior (before @chrisfilo's changes) would be the implementation for `force_nifti_method_2 == False`

In summary:
- method 1 would not be supported and replaced with a method 2 solution from the available info (qform and sform).
- method 2 will be generally enforced (default `force_nifti_method_2` would be `True`).
- method 3 would be available (and would work with the heuristics currently implemented in fmriprep).
- An `afni_compat` would ensure that sform and qform are the same.

**Comment by effigies** [comment]
Apologies if this gets a little ranty, but I think this proposal heads in exactly the wrong direction. The [Recommended usage of qform and sform](https://nifti.nimh.nih.gov/nifti-1/documentation/nifti1fields/nifti1fields_pages/qsform_brief_usage) guide is, in my opinion, more of a historical document than a normative one, and I think there are good reasons not to follow it.

The premise of the document is as follows:

>  The `qform` and `sform` stored in a nifti file are intended to fulfil the following functions:
> * specify the handedness of the coordinate system (important for getting left and right correct)
> * specify the original scanner coordinates (`qform` only; figure)
> * specify standard space coordinates (`sform` only)
> * specify a relationship with another image's coordinates (`sform` only) 

While I agree that this is the intention, what this means is that `qform` should only change if I reorder the axes, or resample the data so that the data array is now different. Meanwhile the `sform` can change however is needed to make the image interpretable with relation to some other image. Therefore these matrices must be allowed to get out-of-sync.

As we've discovered, out-of-sync xforms lead to different behavior in different toolkits, e.g. FSL and ANTs. We've instead opted for a pattern of coding the xforms so that they are always in sync, to ensure consistent behavior without constantly having to refer back to an "original" image to translate transforms between ITK-compatible and FSL-compatible versions. To follow the guidelines in the linked document is to completely change course after a *lot* of work hunting down bug after bug that boiled down to affine mismatches, whether in the image being transformed or the reference image being used to convert between ITK and FSL.

Therefore, I would completely ignore the requirements that `qform_code in (0, 1)` and `sform_code in (0, 2, 3, 4)`. I think we should only consider 1 (scanner), 2 (aligned), and 4 (MNI), for the reason that 0 (unknown) is an error and we never use Talairach (3) space. I would interpret 1 as "unaligned to any reference", because we don't care about scanner space at all.

As to the more specific proposals, I haven't fully digested them, yet. I'm not sure what they buy us, but I'll have another look later.

**Comment by oesteban** [comment]
Should we narrow this down to just incorporating two more checks into SanitizeIMage?:
- [ ] Check units. Check correct units if missing (and write so in the output report).
- [ ] Check RepetitionTime. Confirm nifti header and sidecar json match, rewrite nifti header when they don't match. Report if something was fixed, making it clear that the json sidecar was used over the nifti header.

**Comment by oesteban** [comment]
In that case, the only reason not to close this issue without further action is whether we want to catch users abusing the `--skip-bids-validator` argument. I'll close it for now then, please reopen if we want to implement these "police" behaviors.
